### PR TITLE
Upgrade to OpenIddict 4.0 preview.

### LINKED ISSUE #14483: Upgrade to OpenIddict 4.0
* We will try to migrate to OpenIddict 4.0-preview4, in a separate branch, and see if it is properly working. If it is working as we expected, we will ship ABP 7.0-RC.1 with OpenIddict 4.0 preview, on November 15.
* Once the final release of OpenIddict 4.0 published, we will immediately upgrade and test it with ABP 7.0. So, ABP 7.0 final (that will probably be released on December 13) will use the OpenIddict 4.0 final release.

### PR DESCRIPTION
**Pull request by maliming**
Resolve #14483

### CODE REVIEW

**Review by hikalkan — APPROVED** [review]

File: modules/openiddict/src/Volo.Abp.OpenIddict.AspNetCore/Volo/Abp/OpenIddict/Handlers/AttachAbpCustomChallengeErrors.cs [hunk 1]
```diff
@@ -26,13 +24,9 @@ public ValueTask HandleAsync(OpenIddictServerEvents.ProcessChallengeContext cont
     {
         Check.NotNull(context, nameof(context));
 
-        var properties = context.Transaction.Properties[typeof(AuthenticationProperties).FullName!].As<AuthenticationProperties>();
```
Comment (kevinchalet) [inline modules/openiddict/src/Volo.Abp.OpenIddict.AspNetCore/Volo/Abp/OpenIddict/Handlers/AttachAbpCustomChallengeErrors.cs:29]: FYI, there's a built-in handler in both OpenIddict 3.x and 4.x that already does that for the custom parameters added to `AuthenticationProperties.Parameters` so this handler shouldn't be necessary (and the cool thing is that you can return strongly-typed JSON properties via `AuthenticationProperties.Parameters`) 😃

**Comment by kevinchalet** [comment]
In case you'd be interested, I plan to release another 4.0 preview this weekend (or next week) that will target ASP.NET Core 7.0 RC2 and include the workaround I mentioned in our mail exchange.

Have a nice weekend!

File: modules/openiddict/src/Volo.Abp.OpenIddict.AspNetCore/Volo/Abp/OpenIddict/Handlers/AttachAbpCustomChallengeErrors.cs (hunk 1 above)
Comment (maliming) [reply modules/openiddict/src/Volo.Abp.OpenIddict.AspNetCore/Volo/Abp/OpenIddict/Handlers/AttachAbpCustomChallengeErrors.cs:29]: hi @kevinchalet 

I didn't find such a handler in 3.x.

The custom parameters are from `AuthenticationProperties`.

```cs
var properties = new AuthenticationProperties(new Dictionary<string, string>
{
    [OpenIddictServerAspNetCoreConstants.Properties.Error] = OpenIddictConstants.Errors.InvalidGrant,
    [OpenIddictServerAspNetCoreConstants.Properties.ErrorDescription] = nameof(SignInResult.RequiresTwoFactor),
    ["userId"] = Guid.NewGuid().ToString("N"),
    ["twoFactorToken"] = Guid.NewGuid().ToString("N")
});
return Forbid(properties, OpenIddictServerAspNetCoreDefaults.AuthenticationScheme);
```

Comment (kevinchalet) [reply modules/openiddict/src/Volo.Abp.OpenIddict.AspNetCore/Volo/Abp/OpenIddict/Handlers/AttachAbpCustomChallengeErrors.cs:29]: @maliming hey 😄 

`AuthenticationProperties` has two collections:
  - `AuthenticationProperties.Items` (which was called `AuthenticationProperties.Dictionary` in OWIN/Katana): it's an untyped collection that allows flowing specific properties between the user code and the handler. Some handlers persist them in things like cookies or tokens, some don't (in OpenIddict 4.x, we'll bring back the logic that persisted these properties in the tokens as in OpenIddict 1.x/2.x).

  - `AuthenticationProperties.Parameters`: it was introduced in ASP.NET Core 2.1 and is of type `IDictionary<string,object?>`. It's the one you need to use to return custom parameters via the built-in handler.

TL;DR, you'll basically want to use the newer constructor that takes 2 dictionaries:

```csharp
var properties = new AuthenticationProperties(
    items: new Dictionary<string, string>
    {
        [OpenIddictServerAspNetCoreConstants.Properties.Error] = OpenIddictConstants.Errors.InvalidGrant,
        [OpenIddictServerAspNetCoreConstants.Properties.ErrorDescription] = nameof(SignInResult.RequiresTwoFactor)
    },
    parameters: new Dictionary<string, object>
    {
        ["userId"] = Guid.NewGuid().ToString("N"),
        ["twoFactorToken"] = Guid.NewGuid().ToString("N")
    });

return Forbid(properties, OpenIddictServerAspNetCoreDefaults.AuthenticationScheme);
```

Comment (maliming) [reply modules/openiddict/src/Volo.Abp.OpenIddict.AspNetCore/Volo/Abp/OpenIddict/Handlers/AttachAbpCustomChallengeErrors.cs:29]: Thanks @kevinchalet 

I got that. 💯

**Comment by kevinchalet** [comment]
FYI OpenIddict 4.0 preview5 just shipped (and it targets the final .NET 7.0 RTM packages): https://github.com/openiddict/openiddict-core/releases/tag/4.0.0-preview5 🎉
